### PR TITLE
SHIELD-229-make-cert-rotate-idempotent

### PR DESCRIPTION
**Pull request by prasad927**
Signed-off-by: prasad927 <Prasad.Chaudhari@progress.com>

### :chains: Related Resources
https://chefio.atlassian.net/browse/SHIELD-229

### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [ ] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [ ] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [ ] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [ ] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [ ] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/ES634qGMPVdNkv0GWobHu1sBzWC25gfBCn_uEEZLa3fiJg?e=J1iVYR

https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/Ebnpo0P4nzxPh663KgIjBcEBCnJhQHg17Qf_uqUZC3LZaQ?e=Jqly8L

https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/EUh_OJHkqkBFr8PHlGQVkagBBSscyUGVt-DJQs-PlKxuoA?e=QsX8uB

https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/ERnaeJCeERVNiYaeXTXMPt4BrmNEBrUEU2veNoPu8eIPhg?e=NfcQMw

https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/ETrEKCDzOzNOhNz6IOwyBpkBaeRicdAlttrv6m7tmt7BcA?e=o4lUVg

https://progresssoftware.sharepoint.com/:i:/s/ChefCoreC/Ec6TOiLhvpJCoKz6Y7N5_GgBavaWxQZ7f6yitaa5Eq9OCQ?e=PYgMe3

### CODE REVIEW

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 1]
```diff
@@ -486,6 +528,134 @@ func (c *certRotateFlow) getAllIPs(infra *AutomteHAInfraDetails) []string {
...
+	}
+	return skipIpsList
+}
+
+//comparePublicCertAndPrivateCert compare new public-cert and private cert with current public-cert and private-cert and returns ips to skip cert-rotation.
+func (c *certRotateFlow) comparePublicCertAndPrivateCert(newCerts *certificates, certByIpList []CertByIP, isCertsSame bool, flagsObj *certRotateFlags) []string {
+	skipIpsList := []string{}
+	for _, currentCerts := range certByIpList {
+		isCertsSameTemp := false
+		if strings.TrimSpace(currentCerts.PublicKey) == newCerts.publicCert && strings.TrimSpace(currentCerts.PrivateKey) == newCerts.privateCert && isCertsSame {
```
Comment (atultherajput) [inline components/automate-cli/cmd/chef-automate/certRotate.go:584]: Add () for string comparison.
`if (strings.TrimSpace(currentCerts.PublicKey) == newCerts.publicCert) && (strings.TrimSpace(currentCerts.PrivateKey) == newCerts.privateCert) && isCertsSame {`

File: components/automate-cli/cmd/chef-automate/ha_cert_show.go [hunk 2]
```diff
@@ -114,26 +124,26 @@ func (c *certShowImpl) certShow(cmd *cobra.Command, args []string) error {
...
-			return err
+			return nil, err
 		}
 	} else {
-		return status.New(status.ConfigError, "Invalid deployer type. Either architecture.existing_infra or architecture.aws must be set in config.toml")
+		return nil, status.New(status.ConfigError, "Invalid deployer type. Either architecture.existing_infra or architecture.aws must be set in config.toml")
 	}
 
-	return c.validateAndPrintCertificates(remoteService, certInfo)
+	return certInfo, nil
```
Comment (shaik80) [inline components/automate-cli/cmd/chef-automate/ha_cert_show.go:146]: why we have removed `validateAndPrintCertificates`

Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/ha_cert_show.go:146]: its not removed its just shifted.

**Review by shaik80 — APPROVED** [review]

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 3]
```diff
@@ -93,6 +94,13 @@ const (
 	sudo systemctl start hab-sup.service`
 
 	IP_V4_REGEX = `(\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`
+
+	ERROR_SELF_MANAGED_DB_CERT_ROTATE = "Certificate rotation for externally configured %s is not supported."
+	SKIP_IPS_MSG_CERT_ROTATE          = "The following %s %s will skip while certificate rotation as the following %s have the same certificates as currently provided certificates.\n\t %s"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/certRotate.go:99]:
```suggestion
	SKIP_IPS_MSG_CERT_ROTATE          = "The following %s %s will skip during certificate rotation as the following %s have the same certificates as currently provided certificates.\n\t %s"
```

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 4]
```diff
@@ -93,6 +94,13 @@ const (
...
+	SKIP_FRONT_END_IPS_MSG_A2         = "The following %s %s will skip while root-ca patching as the following %s have same root-ca as currently provided automate root-ca.\n\t %s"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/certRotate.go:100]:
```suggestion
	SKIP_FRONT_END_IPS_MSG_A2         = "The following %s %s will skip during root-ca patching as the following %s have same root-ca as currently provided automate root-ca.\n\t %s"
```

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 5]
```diff
@@ -93,6 +94,13 @@ const (
...
+	SKIP_FRONT_END_IPS_MSG_PG         = "The following %s %s will skip while root-ca patching as the following %s have same root-ca as currently provided postgres root-ca.\n\t %s"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/certRotate.go:101]:
```suggestion
	SKIP_FRONT_END_IPS_MSG_PG         = "The following %s %s will skip during root-ca patching as the following %s have same root-ca as currently provided postgres root-ca.\n\t %s"
```

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 6]
```diff
@@ -93,6 +94,13 @@ const (
...
+	SKIP_FRONT_END_IPS_MSG_OS         = "The following %s %s will skip while root-ca and common name patching as the following %s have same root-ca and common name as currently provided opensearch root-ca and common name.\n\t %s"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/certRotate.go:102]:
```suggestion
	SKIP_FRONT_END_IPS_MSG_OS         = "The following %s %s will skip during root-ca and common name patching as the following %s have same root-ca and common name as currently provided OpenSearch root-ca and common name.\n\t %s"
```

File: components/automate-cli/cmd/chef-automate/certRotate.go [hunk 7]
```diff
@@ -93,6 +94,13 @@ const (
...
+	SKIP_FRONT_END_IPS_MSG_CN         = "The following %s %s will skip while common name patching as the following %s have same common name as currently provided.\n\t %s"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/certRotate.go:103]:
```suggestion
	SKIP_FRONT_END_IPS_MSG_CN         = "The following %s %s will skip during common name patching as the following %s have same common name as currently provided.\n\t %s"
```

File: components/automate-cli/cmd/chef-automate/certRotate.go (hunk 3 above)
Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/certRotate.go]: resolved

File: components/automate-cli/cmd/chef-automate/certRotate.go (hunk 4 above)
Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/certRotate.go]: resolved

File: components/automate-cli/cmd/chef-automate/certRotate.go (hunk 5 above)
Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/certRotate.go]: resolve

File: components/automate-cli/cmd/chef-automate/certRotate.go (hunk 6 above)
Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/certRotate.go]: resolve

File: components/automate-cli/cmd/chef-automate/certRotate.go (hunk 7 above)
Comment (prasad927) [reply components/automate-cli/cmd/chef-automate/certRotate.go]: resolve

**Review by atultherajput — APPROVED** [review]

**Review by vivekshankar1 — APPROVED** [review]

**Review by SanjuPal01 — APPROVED** [review]

**Review by vivek-yadav — APPROVED** [review]

**Pull request merged**